src/Data/Tree/AVL/Map/Membership.agda

```diff
@@ -0,0 +1,159 @@
+
+------------------------------------------------------------------------
+-- Membership: ∈
+
+_∈_ : Key × V → Map V → Set _
+kx ∈ m = Any (Is kx) m
+
+_∉_ : Key × V → Map V → Set _
+kx ∉ m = ¬ kx ∈ m
+
+∈-Respectsˡ : _∈_ {V = V} Respectsˡ Is
+∈-Respectsˡ x~y = Any.map (Is-trans (Is-sym x~y))
+
+∈-empty : kx ∉ empty
+∈-empty (tree ())
+
+------------------------------------------------------------------------
+-- singleton
+
+∈-singleton⁺ : (k , x) ∈ singleton k x
+∈-singleton⁺ = tree (IAnyₚ.singleton⁺ _ _ (refl , ≡-refl))
+
+∈-singleton⁻ : (k , x) ∈ singleton k′ x′ → k ≈ k′ × x ≡ x′
+∈-singleton⁻ (tree p) = IAnyₚ.singleton⁻ _ _ p
+
+private
+  ≈-lookup : (p : (k , x) ∈ m) → k ≈ Any.lookupKey p
+  ≈-lookup (tree p) = proj₁ (IAnyₚ.lookup-result p)
+
+------------------------------------------------------------------------
+-- insert
+
+∈-insert⁺ : k ≉ k′ → (k , x) ∈ m → (k , x) ∈ insert k′ x′ m
+∈-insert⁺ {k′ = k′} {m = m@(tree t)} k≉k′ (tree p) =
+  tree (IAnyₚ.insert⁺ _ _ _ (⊥⁺<[ _ ]<⊤⁺) p k′≉key-p)
+  where
+  k′≉key-p : k′ ≉ Any.lookupKey (tree p)
+  k′≉key-p k′≈key-p = k≉k′ (Eq.trans (≈-lookup (tree p)) (Eq.sym k′≈key-p))
+
+∈-insert⁺⁺ : (k , x) ∈ insert k x m
+∈-insert⁺⁺ {k = k} {m = tree t} with IAny.any? ((k ≟_) ∘ key) t
+... | yes k∈ = tree (IAnyₚ.Any-insert-just _ _ _ _ (λ k′ → _, ≡-refl) k∈)
+... | no ¬k∈ = tree (IAnyₚ.Any-insert-nothing _ _ _ _ (refl , ≡-refl) ¬k∈)
+
+private
+  Is-Resp : ∀ {kv k k′ v} → k ≈ k′ → Is {V = V} kv (k′ , v) → Is kv (k , v)
+  Is-Resp = (λ{ k′≈l (k≈l , x≡) → (trans k≈l (sym k′≈l) , x≡)})
+
+∈-insert⁻ : (k , x) ∈ insert k′ x′ m → (k ≈ k′ × x ≡ x′) ⊎ (k ≉ k′ × (k , x) ∈ m)
+∈-insert⁻ {m = tree t} (tree kx∈insert)
+    with IAnyₚ.insert⁻ Is-Resp _ _ t (⊥⁺<[ _ ]<⊤⁺) kx∈insert
+... | inj₁ p = inj₁ p
+... | inj₂ p = let k′≉p , k≈p , _ = IAnyₚ.lookup-result p
+                   k≉k′ = λ k≈k′ → k′≉p (trans (sym k≈k′) k≈p)
+               in inj₂ (k≉k′ , tree (IAny.map proj₂ p))
+
+------------------------------------------------------------------------
+-- lookup
+
+∈-lookup⁺ : (k , x) ∈ m → lookup m k ≡ just x
+∈-lookup⁺ {k = k} {m = tree t} (tree kx∈m)
+    with IAnyₚ.lookup⁺ t k (⊥⁺<[ _ ]<⊤⁺) kx∈m | IAnyₚ.lookup-result kx∈m
+... | inj₁ p≉k        | k≈p , x≡p = contradiction (sym k≈p) p≉k
+... | inj₂ (p≈k , eq) | k≈p , x≡p = ≡-trans eq (cong just (≡-sym x≡p))
+
+∈-lookup⁻ : lookup m k ≡ just x → (k , x) ∈ m
+∈-lookup⁻ {m = tree t} {k = k} {x = x} eq
+  = tree (IAny.map (Prod.map sym ≡-sym) (IAnyₚ.lookup⁻ t k x (⊥⁺<[ _ ]<⊤⁺) eq))
+
+∈-lookup-nothing⁺ : (∀ x → (k , x) ∉ m) → lookup m k ≡ nothing
+∈-lookup-nothing⁺ {k = k} {m = m@(tree t)} k∉m with lookup m k in eq
+... | nothing = ≡-refl
+... | just x = contradiction (∈-lookup⁻ eq) (k∉m x)
+
+∈-lookup-nothing⁻ : lookup m k ≡ nothing → (k , x) ∉ m
+∈-lookup-nothing⁻ eq kx∈m with ≡-trans (≡-sym eq) (∈-lookup⁺ kx∈m)
+... | ()
+
+------------------------------------------------------------------------
+-- ∈?
+
+∈-∈? : (k , x) ∈ m → (k ∈? m) ≡ true
```

Comment:
Okay so here we run into a little bit of a naming problem. The name `_∈?_` is usually reserved for the full proof relevant decision procedure, rather than the boolean procedure which it currently is. Obviously this is a problem with existing code, not your additions. Nonetheless it does affect the naming of your new lemmas!

I'd like to make room for (eventually!) adding the proof relevant decision procedure. I think the best option would be for you to deprecate the name ` _∈?_` in the original file and replace it with the name `member`? Would you be willing to do that?